presentations/vocabulary_types_2023_11_8.md

```diff
@@ -0,0 +1,241 @@
+| An instance of one of a closed-set of types `Ts...`| `std::variant<Ts...>`|
+| `N` instances of a type `T`| `std::array<T, N>`|
+| Variable-count, multiple instances of a type `T`| `std::vector<T>`|
+| Unique, variable-count, instances of a type `T`| `std::set<T>`|
+| Key-accessed, instances of a type `T`| `std::map<Key, T>`|
+
+---
+
+# Special member functions
+
+The compiler can generate special member functions. 
+Each special member function can be compiler-generated if it is supported by all component objects.
+
+| Special member(s) | Signature(s) |
+|----|---|
+| Default constructor | `T();` |
+| Destructor | `~T();` |
+| Copy constructor/assignment | `T(const T&);` `T& operator=(const T&);` |
+| Move constructor/assignment|  `T(T&&);` `T& operator=(T&&);` |
+
+---
+
+# Const propagation
+
+The compiler will only allow const-qualified member functions to be called 
+on components when they are accessed through a const-access path.
+
+We call this _const-propagation_.
+
+Standard library vocabulary types provide const-qualified and non-const-qualified overloads of accessors to owned objects to enforce const propagation:
+
+```c++
+const T& std::vector<T>::operator[](size_t index) const; 
+T& std::vector<T>::operator[](size_t index); 
+```
+
+---
+
+# Requirements for vocabulary types
+
+Composite classes built with vocabulary types should be composable:
+
+* const propagates.
+
+* The compiler can generate all special member functions where they are defined for an owned type `T`.
+
+These requirements make working with C++ easier - classes should do business logic or resource management, not both.
+
+---
+
+# Combining vocabulary types
+
+We can combine vocabulary types to express combined idioms:
+
+| Idiom | Type |
+|---|---|
+| Variable-count, multiple, nullable instances of a type `T`| `std::vector<std::optional<T>>`|
+| Key-accessed, instances of one of a closed-set of types `Ts...`| `std::map<Key, std::variant<Ts...>>`|
+
+---
+
+# Incomplete types
+
+We may want a composite to contain an instance of an incomplete type, either directly or through the use of existing vocabulary types.
+
+Incomplete type support is needed for defining recursive data structures, supporting open-set polymorphism, and hiding implementation detail.
+
+Incomplete types are supported by storing the object of the incomplete type outside of the composite object.
+
+Storing an object outside of a composite type can also be used to optimize use of cache (hot-cold-splitting).
+
+---
+
+## Incomplete types using pointers
+
+Since we want to store the object of incomplete type outside of our composite type,
+pointers are the first thing we may reach for:
+
+```c++
+class Composite {
+    T* ptr_;
+    // ...
+}
+```
+
+Pointers are a poor fit for owned data, we'll need to implement all special member functions and manually check const-qualified methods - const does not propagate to a pointee when the pointer is accessed through a const-access path.
```

Comment:
Pointers are a poor fit for owned data as we need to implement all special member functions and manually check const-qualified methods (const does not propagate to a pointee when the pointer is accessed through a const-access path).